Interface conversion: composition

The following program `sample.go` triggers a panic:

```go
package main

import "fmt"

type Root struct {
	Name string
}

func (r *Root) Hello() {
	fmt.Println("Hello", r.Name)
}

type One struct {
	Root
}
type Hi interface {
	Hello()
}

func main() {
	var root interface{} = &Root{Name: "test"}
	var one interface{} = &One{Root{Name: "test"}}

	fmt.Println(root.(Hi))
	one.(Hi).Hello()
}

// Output:
// &{test}
// Hello test

```

Expected result:
```console
$ go run ./sample.go
&{test}
Hello test
```

Got:
```console
$ dyngo ./sample.go
panic: interface conversion: interface {} is *struct { Name string }, not interp.valueInterface [recovered]
	panic: interface conversion: interface {} is *struct { Name string }, not interp.valueInterface

goroutine 1 [running]:
github.com/containous/dyngo/interp.runCfg.func1(0xc000186640)
	/home/ldez/sources/go/src/github.com/containous/dyngo/interp/run.go:103 +0x134
panic(0xbef4a0, 0xc0001c7b30)
	/home/ldez/.gvm/gos/go1.11.5/src/runtime/panic.go:513 +0x1b9
github.com/containous/dyngo/interp.genValueInterfaceValue.func1(0xc000186640, 0x1, 0x1, 0xc00017b060)
	/home/ldez/sources/go/src/github.com/containous/dyngo/interp/value.go:108 +0xaa
github.com/containous/dyngo/interp.callBin.func4(0xc000186640, 0xc000186550)
	/home/ldez/sources/go/src/github.com/containous/dyngo/interp/run.go:655 +0x100
github.com/containous/dyngo/interp.runCfg(0xc00024a480, 0xc000186640)
	/home/ldez/sources/go/src/github.com/containous/dyngo/interp/run.go:108 +0x62
github.com/containous/dyngo/interp.(*Interpreter).run(0xc0001f4070, 0xc0002498c0, 0xc0001865f0)
	/home/ldez/sources/go/src/github.com/containous/dyngo/interp/run.go:90 +0xa6
github.com/containous/dyngo/interp.(*Interpreter).Eval(0xc0001f4070, 0xc0001f8000, 0x174, 0xca583f, 0x4, 0xc0001f4070, 0x0, 0x115f2e0)
	/home/ldez/sources/go/src/github.com/containous/dyngo/interp/interp.go:236 +0x4c3
main.main()
	/home/ldez/sources/go/src/github.com/containous/dyngo/dyngo.go:50 +0x334
exit status 2

```

--- a/interp/ast.go
+++ b/interp/ast.go
@@ -209,6 +209,7 @@ const (
 	Lor
 	Lower
 	LowerEqual
+	Method
 	Mul
 	MulAssign
 	Negate
@@ -263,6 +264,7 @@ var actions = [...]string{
 	Land:         "&&",
 	Lor:          "||",
 	Lower:        "<",
+	Method:       "Method",
 	Mul:          "*",
 	MulAssign:    "*=",
 	Negate:       "-",

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -956,7 +956,9 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 				n.val = ti
 				switch n.typ.cat {
 				case InterfaceT:
+					n.typ = n.typ.fieldSeq(ti)
 					n.gen = getMethodByName
+					n.action = Method
 				case PtrT:
 					n.typ = n.typ.fieldSeq(ti)
 					n.gen = getPtrIndexSeq

--- a/interp/interp_eval_test.go
+++ b/interp/interp_eval_test.go
@@ -250,6 +250,38 @@ func TestEvalUnary(t *testing.T) {
 	})
 }
 
+func TestEvalMethod(t *testing.T) {
+	i := interp.New(interp.Opt{})
+	eval(t, i, `
+		type Root struct {
+			Name string
+		}
+
+		type One struct {
+			Root
+		}
+
+		type Hi interface {
+			Hello() string
+		}
+
+		func (r *Root) Hello() string { return "Hello " + r.Name }
+
+		var r = Root{"R"}
+		var o = One{r}
+		var root interface{} = &Root{Name: "test1"}
+		var one interface{} = &One{Root{Name: "test2"}}
+	`)
+	runTests(t, i, []testCase{
+		{src: "r.Hello()", res: "Hello R"},
+		{src: "(&r).Hello()", res: "Hello R"},
+		{src: "o.Hello()", res: "Hello R"},
+		{src: "(&o).Hello()", res: "Hello R"},
+		{src: "root.(Hi).Hello()", res: "Hello test1"},
+		{src: "one.(Hi).Hello()", res: "Hello test2"},
+	})
+}
+
 func runTests(t *testing.T, i *interp.Interpreter, tests []testCase) {
 	for _, test := range tests {
 		t.Run(test.desc, func(t *testing.T) {

--- a/interp/run.go
+++ b/interp/run.go
@@ -114,12 +114,19 @@ func typeAssert(n *Node) {
 	i := n.findex
 	next := getExec(n.tnext)
 
-	if n.child[0].typ.cat == ValueT {
+	switch {
+	case n.child[0].typ.cat == ValueT:
 		n.exec = func(f *Frame) Builtin {
 			f.data[i] = value(f).Elem()
 			return next
 		}
-	} else {
+	case n.child[1].typ.cat == InterfaceT:
+		n.exec = func(f *Frame) Builtin {
+			v := value(f).Interface().(valueInterface)
+			f.data[i] = reflect.ValueOf(valueInterface{v.node, v.value})
+			return next
+		}
+	default:
 		n.exec = func(f *Frame) Builtin {
 			v := value(f).Interface().(valueInterface)
 			f.data[i] = v.value
@@ -134,7 +141,8 @@ func typeAssert2(n *Node) {
 	value1 := genValue(n.anc.child[1]) // returned status
 	next := getExec(n.tnext)
 
-	if n.child[0].typ.cat == ValueT {
+	switch {
+	case n.child[0].typ.cat == ValueT:
 		n.exec = func(f *Frame) Builtin {
 			if value(f).IsValid() && !value(f).IsNil() {
 				value0(f).Set(value(f).Elem())
@@ -142,7 +150,14 @@ func typeAssert2(n *Node) {
 			value1(f).SetBool(true)
 			return next
 		}
-	} else {
+	case n.child[1].typ.cat == InterfaceT:
+		n.exec = func(f *Frame) Builtin {
+			v, ok := value(f).Interface().(valueInterface)
+			value0(f).Set(reflect.ValueOf(valueInterface{v.node, v.value}))
+			value1(f).SetBool(ok)
+			return next
+		}
+	default:
 		n.exec = func(f *Frame) Builtin {
 			v, ok := value(f).Interface().(valueInterface)
 			value0(f).Set(v.value)
@@ -453,8 +468,7 @@ func call(n *Node) {
 		// Compute method receiver value
 		values = append(values, genValueRecv(n.child[0]))
 		method = true
-	}
-	if n.child[0].typ.cat == InterfaceT {
+	} else if n.child[0].action == Method {
 		// add a place holder for interface method receiver
 		values = append(values, nil)
 		method = true
@@ -529,7 +543,11 @@ func call(n *Node) {
 				if v == nil {
 					src = def.recv.val
 					if len(def.recv.index) > 0 {
-						src = src.FieldByIndex(def.recv.index)
+						if src.Kind() == reflect.Ptr {
+							src = src.Elem().FieldByIndex(def.recv.index)
+						} else {
+							src = src.FieldByIndex(def.recv.index)
+						}
 					}
 				} else {
 					src = v(f)

